### ISSUE TITLE
[FEA] Try UVM in RMM to reduce cugraph's memory requirement on SNSG

### ISSUE THREAD

**Issue by afender** [rapidsai/cugraph#151]
**Is your feature request related to a problem? Please describe.**
We identified that cugraph’s peak memory requirement is in coo2csr. The current implementation sorts the edge list and then builds the CSR matrix

**Describe the solution you'd like**
RMM now supports UVM. This would solve the problem of the GPU memory limit by using oversubscription. Performance-wise we should expect to see a performance hit (for largest sizes only). 

**Describe alternatives you've considered**
Sort-free algorithm (also being investigated)
Manual memory management + sort that can stream data (much more complex to develop)
Use Multiple GPUs (long term solution - need something sooner)

**Task List**
- [x] enable UVM : it should just be a RMM parameter to set
- [ ] get memory profile and performance results on large datasets (>1B edges)

**Additional context**
In cuGraph 0.7 we will also have renumbering which will add another set of COO and increase the memory requirement further. This will not only help with coo2csr but also all places where memory requirements are high.

**Comment by afender** [comment]
@seunghwak at this point you are not planning on adding code to the repo as part of this experiment, correct?

**Comment by seunghwak** [comment]
Correct!

**Comment by afender** [comment]
@

**Comment by seunghwak** [comment]
There are two different ways to use unified memory in cuGraph.

One is to enable unified memory without RMM's pool allocation and another is to enable unified memory with RMM's pool allocation.

When RMM is configured to use both pool allocation and unified memory, RMM allocates the initial pool using cudaMallocManaged and prefetches the pool to the GPU using cudaMemPrefetchAsync.

So in the first case (no pool allocation), memory pages are created when first accessed in COOtoCSR; but in the second case (with pool allocation), memory pages are created in RMM initialization.

So, COOtoCSR runs noticeably slower (more than 3X) when unified memory is enabled without pool allocation, but the performance difference is hardly noticeable when unified memory is enabled with pool allocation (as memory page creation overhead is already included in RMM initialization) as long as the memory footprint is within the limit of the initial RMM pool size.

The following gives COOtoCSR execution times with different RMM configurations (HiBench pagerank input data with 12M pages).

managed: True pool: True
=> 0.233247 s
managed: True pool: False
=> 0.790841 s
managed: False pool: True
=> 0.233355 s
managed: False pool: False
=> 0.26419 s

cnmem does not allow oversubscription so I cannot enable RMM's pool allocation with oversubscription; see cnmem.cpp's cnmemInit function.

```cpp
1103         CNMEM_CHECK_TRUE( 
1104             size > 0 && size < props.totalGlobalMem, CNMEM_STATUS_INVALID_ARGUMENT);
```
=> if the asked size exceeds the available physical GPU memory, cnmemInit fails.

So I can run only with pool set to False.

For HiBench pagerank input data with 25M pages, 

COOtoCSR took 2.07944 s with managed set to true, while COOtoCSR fails if managed is set to false (terminate called after throwing an instance of 'thrust::system::system_error'
  what():  rmm_allocator::allocate(): RMM_ALLOC: unspecified launch failure).

Considering that the data size is only slightly larger than 2x, this is a significant slow down. We may be able to improve performance with manual hints (e.g. cudaMemAdvise and cudaMemPrefetchAsync, for example, we do not need to keep input COO data in the GPU memory once we created a copy), but I deferred this task as we are updating the COOtoCSR algorithm.

All the experiments are ran on GV100 with 32 GB memory, CUDA 10.0, and Ubuntu 18.0.4.
